WalletWasabi/Extensions/RPCClientExtensions.cs

```diff
@@ -114,6 +124,61 @@ public static async Task<AllFeeEstimate> EstimateAllFeeAsync(this IRPCClient rpc
 				.ToDictionary(x => x.target, x => (int)Math.Ceiling(x.feeRate.SatoshiPerByte));
 		}
 
+		private static Dictionary<int, int> GetFeeEstimationsFromMempoolInfo(MemPoolInfo mempoolInfo)
+		{
+			const int BlockSize = 1_000_000;
+			static IEnumerable<(int Size, FeeRate From, FeeRate To)> SplitFeeGroupInBlocks(FeeRateGroup group)
+			{
+				var remainingSize = (long)group.Sizes;
+				while (remainingSize > 0)
+				{
+					var chunckSize = (int)Math.Min(remainingSize, BlockSize);
+					yield return (Size: chunckSize, From: group.From, To: group.To);
+					remainingSize -= BlockSize;
+				}
+			}
+			// Filter those groups with very high fee transactions (less than 0.1%).
+			// This is because in case a few transactions pay unreasonablely high fees
+			// then we don't want our estimations to be affected by those rare cases.
+			var relevantFeeGroups = mempoolInfo.Histogram
+				.OrderByDescending(x => x.Group)
+				.SkipWhile(x => x.Count < mempoolInfo.Size / 1_000);
+
+			// Splits multi-megabyte fee rate groups in 1mb chunck
+			// We need to count blocks (or 1MvB transaction chuncks) so, in case fee
+			// groups are bigger than 1MvB we split those in multiple 1MvB chuncks.
+			var splittedFeeGroups = relevantFeeGroups.SelectMany(x => SplitFeeGroupInBlocks(x));
+
+			// Assigns the corresponding confirmation target to the set of fee groups.
+			// We have multiple fee rate groups which size are in the range [0..1MvB)
+			//
+			// Example: imagine we have only 4 fee rate groups in the form (size, from, to)
+			//      [(10kb, 400, 500) (55kb, 300, 400) (310kb, 200, 300) (700kb, 100, 200)]
+			//
+			// In this case the three first fee rate groups fit well in the next block so
+			// they have target=1 while the fourth will need to wait and for that reason it
+			// is target=2
+			var accumulatedSizes = splittedFeeGroups
+				.Select(x => x.Size) 
+				.Scan(0m, (acc, size) => acc + size);
+
+			var feeGroupsByTarget = splittedFeeGroups.Zip(accumulatedSizes, (feeGroup, accumulatedSize) =>
+				(FeeRate: feeGroup.From, Target: (int)Math.Ceiling(1 + accumulatedSize / BlockSize)));
```

Comment:
Why `feeGroup.From` is used (given that descending ordering is used)? I would kind of expect `feeGroup.To`.

Suppose that the following holds: `[(5MvB, 400, 500) (55kb, 300, 400)]` then it's more precise to use fee `500` than `400`, isn't it?

Maybe I'm missing something so take this with a grain of salt.

Reply:
Using `To` is also possible and deciding which is the best one depends on what one wants to get. In this case I wanted to get the minimum fee rate required to confirm in a given block so, imagine in the top of the mempool there are 5MvB transactions paying between 400s/vb and 500s/vb (`[(5MvB, 400, 500)]`) what is the minimum fee rate we have to pay in order to be included in that range? 

* 400s/vb is the absolute minimum but that does not guarantee anything really,
* 500s/vb is the maximum paid in the range what means if a tx pays more it will be in that segment by sure.

I went with the minimum because I saw it provides the best estimation (empirically observed) for the next 3 blocks (targets=1,2,3) however, if I would take `To` it would have overestimated the fee rates for at least the first target (1).

It is also possible to use `From` for the *economical mode* and `To` for the *conservative mode* too.

Reply:
Thanks for explaining this. 

> It is also possible to use `From` for the _economical mode_ and `To` for the _conservative mode_ too.

That sounds like a good idea.